### DISCUSSION TITLE
alternate backends (brew, nix, npm, cargo, etc)

### DISCUSSION THREAD

**Discussion by jdx** [jdx/mise#1250] [General]
One of the most common feature requests for a long time is to support being able to directly install tools without having an asdf plugin for it. As an example: `mise use -g cargo-eza` or `mise use -g npm-@antfu/ni` maybe even homebrew?

I'm going to try to use this discussion as a central place to manage my solution which I've started spiking on.

## Supported Backends

These are in rough order of priority:

- [x] cargo via `cargo install`
- [x] npm via `npm i -g`
- [ ] eget (might negate the need of others like `cargo binstall`)
- [ ] swift (someone tell me the best solution here, @pepicrft? I think you sent me something but I forget what it was called, also is eget perhaps an alternative?)
- [x] pip via pipx
- [x] go via `go build` OR https://gobinaries.com/
- [x] cargo via `cargo binstall`
- [ ] node alternatives (pnpm, yarn, bun, etc)
- [ ] pkgx (tea successor)
- [ ] webinstall.dev (seems lacking very much so perhaps not—but does have nice aesthetics I would say)
- [ ] homebrew?
- [ ] nix?

## Syntax

Right now I'm going with a dash to separate the backend from the tool name. I know this isn't great
because you could have tools with dashes in them, but it's really the only delimiter we can use in TOML
without needing to add quotes. I want to avoid something like this:

```toml
[tools]
node = "latest"
"cargo:ripgrep" = "latest"
```

It works fine, just looks ugly.

Alternatives considered:

- `mise use -g cargo:eza`
- `mise use -g cargo/eza`
- `mise use -g cargo#eza`
- `mise use -g cargo!eza`
- `mise use -g cargo|eza`

## Runtime/compiler selection

For v1, I'm simply going to call whichever `cargo` or `npm` is on PATH. That may be something you've defined in your `.mise.toml`. It may be something in your system, just depends on how it is configured. This does mean you will need to manually setup cargo/npm/whatever ahead of time.

I think for v1 that's fine and possibly long-term as well. I think later if we detect the compiler/package manager isn't installed we can prompt to install it.

The runtime selection is a harder problem I think. If I install something with node-18, then run it with node-20, will it work ok? If memory serves, you actually need to run `npm install` again to get it to rebuild the right files (but I think there may have been a slightly easier command). Maybe that's better in more modern node? In any case, I'm jus talking about node here but I could see this being a problem for any scripting language. As you jump between projects the CLIs need to keep working.

I hope not but I may need to put a shim in place to detect if the runtime version changed and if so run whatever command I need to patch it for the current version.

## Storage

Related to the above point is storage. I think these should just be like any asdf-plugin installed thing inside of `~/.local/share/mise/installs/cargo-eza`—note the "cargo-" prefix to prevent the scenario where you have 2 things installed via different tools.

Scoped packages in node may be a problem. I'm not sure how mise will handle `~/.local/share/mise/installs/npm-@antfu/ni`. Perhaps we should slugify that to `~/.local/share/mise/installs/npm-@antfu-ni`. I actually think users will prefer this. I don't think they'll want to have npm CLIs grouped into individual directories.

An outstanding question (my spike will address) is how exactly I install something like this. `npm install --into` is the non-existent (at least I think non-existent) command I think I need.

My biggest question is 

## Questions

- is "backends" the right name? I was also thinking "registry" maybe but leaning towards "backends". I had actually started calling them "plugins" but that turned out to be a horrible idea—cuz they're very much not.

## Existing issues/discussions

- #953
- #686
- #580

**Comment by jdx** [comment]
nice! This actually wasn't too bad: https://github.com/jdx/mise/pull/1452

**Comment by pepicrft** [comment]
>  swift (someone tell me the best solution here, @pepicrft? I think you sent me something but I forget what it was called, also is eget perhaps an alternative?)

Swift Package Manager doesn't have an API that users can leverage to add tasks to their projects. However, it supports adding plugin dependencies, for example [this one](https://github.com/apple/swift-docc-plugin/blob/main/Package.swift#L27) to generate documentation, which extend `swift package` with additional commands. For example `swift package generate-documentation`. We should detect which packages are command plugins, and hook them up.

**Reply by jdx** [reply]
I think you're thinking this is something different, this is installing tools with package managers—nothing to do with tasks.

e.g.:

- `mise install cargo-ripgrep@latest`
- `mise install npm-prettier@latest`
- `mise install swift-<something>@latest`

**Reply by pepicrft** [reply]
Oh, I see what you mean. Surprisingly, there's nothing official. Hence why, the community built [its own tool](https://github.com/yonaskolb/Mint).

**Reply by kattouf** [reply]
[Mint](https://github.com/yonaskolb/Mint) would be great to see here :dreaming:

**Reply by kattouf** [reply]
@jdx I created a [draft-pr](https://github.com/jdx/mise/pull/2199) with [Mint](https://github.com/yonaskolb/Mint) backend and will be glad to receive comments or instructions about next steps.

<img width="650" alt="mise-use" src="https://github.com/jdx/mise/assets/7829589/735e0829-8240-4ab0-a445-8688ecb9c992">

**Reply by pepicrft** [reply]
🎉 this is amazing @kattouf 🎉 . Thanks a lot for going ahead with this.

**Reply by kattouf** [reply]
Hey guys, based on our last conversation in [Mint backend pr](https://github.com/jdx/mise/pull/2199), I'm back with a new "3rd party free" backend implementation.
https://github.com/jdx/mise/pull/2241

**Comment by GeertJohan** [comment]
I would strongly prefer to use a more clear prefix than the dash. e.g. `:`. Even though the TOML becomes a bit ugglier, this indicates to the user that a different installation process is used. Using the prefix with a dash will feel like magic to the user and may lead to a lot of questions.
From a maintenance point of view, when scanning the `.rtx.toml` or `rtx list` output, it's also useful to quickly identify packages that use a plugin vs via package manager.

When considering Go, I'd prefer native `go build/go install` over `gobinaries.com`, although ofcourse both could be implemented (`go:foo@latest` vs `gobinaries:foo@latest`).

Lastly, very go-specific as I'm considering implementing that, Go uses import paths a lot. Where with cargo and npm developers have to 'reserve' their name, with Go it's possible to use any domain name to create an 'import path'. For example `go install github.com/GeertJohan/gomatrix@latest` downloads/compiles and installs a binary named `gomatrix`.

How would we translate this to rtx? A couple of options:
- `go:github.com/GeertJohan/gomatrix@latest`
- `go:github-com-GeertJohan-gomatrix@latest`
- `go:github--com--GeertJohan--gomatrix@latest`

I prefer the first one because it differs the least from original go usage. Need to look into implementation though. TOML will need quoting, but already needs it IF using the `:` as separator for the backend-prefix instead of `-`. The last one, where periods and slashes are replaced with a double dash (`--`) is cleaner when the original import path already has dashes in it. @jdx Any thoughts on this?

**Reply by jdx** [reply]
> From a maintenance point of view, when scanning the .rtx.toml or rtx list output, it's also useful to quickly identify packages that use a plugin vs via package manager.

that's a good point, maybe we'll try the colon for the first pass

**Reply by jdx** [reply]
> I prefer the first one because it differs the least from original go usage. Need to look into implementation though. TOML will need quoting, but already needs it IF using the : as separator for the backend-prefix instead of -. The last one, where periods and slashes are replaced with a double dash (--) is cleaner when the original import path already has dashes in it. @jdx Any thoughts on this?

npm has the same problem with scoped packages I think. I actually didn't really attempt to solve it and mise appears to be fine with tools having a subdirectory.

Whether we want to have it use a subdirectory or not, I suppose that's up to us.

**Reply by jdx** [reply]
so I think colon is a bad choice. If I do that, I won't be able to use a colon in the directories so I'd have to use another separator.

**Reply by jdx** [reply]
that's not a dealbreaker, but I think it's going to make things a lot harder for both me and the users. I already have this problem with `ref:` which gets translated to `ref-`

**Reply by jdx** [reply]
`/` makes for an interesting delimiter. It has a side effect that's actually maybe a good thing—it groups everything together, e.g.:

```sh
mise i npm/@antfu/ni
~/.local/share/mise/installs/npm/@antfu/ni/0.0.0/bin/ni -v
```

I don't know if I love it though, I will admit the CLI syntax of ":" does look better. I just don't like that there are different delimiters in different places.

I also wonder if I should even have the subdirectories. It seems to work fine, but maybe it would be better if I stripped all the subdirectories out and just had a flat directory structure.

**Reply by GeertJohan** [reply]
I think each backend should have it's own directory, and each backend could translate the package name to a directory name. This makes it the most flexible and the core shouldn't have to worry about it. I think it's fine if directory names and package names are not the same, this is 'under the hood', the user rarely sees it (shouldn't have to care) and the backend implementation can have freedom to do the best thing for it's own situation.

I would then also still stick with the colon, the name before the colon is the backend directory (created by the core rtx code) and then the part after the colon is passed to the backend implementation to translate into the directory.

The default 'asdf-plugin' way of working could also be one of these backends by the way, it just has a different install root directory, and it's default (no `asdf-plugin:` prefix required, without a backend prefix the asdf-plugin backend is assumed).

**Reply by GeertJohan** [reply]
e.g.:
```
~/.local/share/mise/installs/asdf-plugin/*       # All existing installs are moved here
~/.local/share/mise/installs/npm/*
~/.local/share/mise/installs/go/*
```

**Comment by RingOfStorms** [comment]
I am most looking forward to `eget` style behavior from _mise_ with the ability to install any binary from releases in git.

Thus solving the pain point of writing asdf plugins which boil down to string manipulating the `uname` command into whatever random naming scheme some project uses for their binaries.

So maybe some process like
`mise git-repo install https://github.com/jdx/mise mise` (this could register the tool name `mise` to use the git-repo releases instead of an asdf plugin)
so that whenever you run `mise install mise@[tag]` it will look for that git Tag and get the binary and use that.

`mise git-repo install https://github.com/zyedidia/eget eget`
`mise global eget@latest` or `mise local eget@v1.3.3` where `v1.3.3` is a git Tag


Or maybe allow for 
`mise install repo_url@tag` and it doesn't care the name of the tool itself.

Note that another challenge may be other files inside of release archives. For example, I want to use mise to install slack's Nebula https://github.com/slackhq/nebula/releases
They release an archive that contains two binaries (which I want both to be installed).

Other repos ship with other files that aren't just the binary files. Idk the best way to solve this some tools similar to eget just look for the largest file in the archive, etc.

So yea to two main issues are:
- Automatically detect current platform (this seems pretty doable based on eget/dra/tool-sync)
- Automatically detect _what_ is actually installable (maybe this is not possible and can be specified for each repo somehow?)

**Reply by RingOfStorms** [reply]
Note other eget like tools built in rust:
- https://github.com/devmatteini/dra
- https://github.com/chshersh/tool-sync/tree/main

**Reply by pepicrft** [reply]
A syntax like `mise install repo_url@tag` would be dope.

This relates somewhat to the issue you brought up about installing multiple files. Some Swift executable binaries might contain dynamic libraries, those need to be downloaded, and some changes to the binary might be necessary to ensure the directory where the dynamic library is used by the dynamic linker to look up binaries.

I couldn't find any convention in eget about this.

**Reply by jdx** [reply]
I think for now it wouldn't be a good use of my time to try to map repos to binaries when so many alternatives are already in this space. I think the better solution is just to be a front-end to those other tools:

```sh
$ mise i egrep:user/repo
$ mise i dra:user/repo
$ mise i pkgx:user/repo
```

**Reply by RingOfStorms** [reply]
I feel like where _mise_ could improve on all of those is being more declarative in the config file or command syntax, to get around those above issues. I was looking at the source code of some of the rust ones above and they fell back to really janky hacks like "Assume the largest file in an archive is the binary and use that one"

Those approaches don't work well and I don't feel like _mise_ will be that great if it just proxies that bad behavior. You'd have to know the ins and out of every single one of these backends which kind of defeats the purpose in my mind. I want to use _mise_ because I don't want to learn all these other backends or do what is required like write asdf-plugins (though maybe that is the sweet spot since that does solve these issues)

**Reply by RingOfStorms** [reply]
I guess something similar to the asdf plugin format but with tools built in and available that make this process easier would be great. The fact that everyone who writes an asdf plugin is grossly copying some `uname -XYZ` commands to determine some string of a binary is really annoying.

If there was just a simplified config for github releases native to mise that was easy to configure and declare that would be great. Where you specify generally what `binaries` exist in the release archives of a git repo. And some code in _mise_ could do the platform specific stuff.

**Reply by jdx** [reply]
sorry I just don't have the appetite to take on this problem and I really think I've broadened the scope of mise plenty enough as it is. If you want a better tool for installing binaries from repos, consider creating one yourself and we could see if it might make sense to adopt it in mise either directly or via a plugin.

**Reply by zph** [reply]
I'm very interested to have a unified approach to installing binaries from Github Releases and I see that there's uncertain appetite for it being in mise directly. I currently use a combination of homebrew and hermit to handle this, but I'm not fully happy with my approach.

I took a few minutes tonight to extend mise to support it via your example @jdx of how `npm` backend was enabled... and I cheated for the prototype by shelling out to `ubi` (https://github.com/houseabsolute/ubi) which handles the release lookup and architecture matching but could be re-implemented or used as a library for a more durable solution.

My prototype implementation was simple to implement thanks to `mise` architecture. 

I'll test drive my own implementation for a bit (https://github.com/zph/mise/tree/zh-add-ubi-forge) and if it holds up well I can polish it up to contribute back.

**Comment by jdx** [comment]
Not sure when I'll be able to get around to doing more of these, if someone wanted to help and had a particular backend they want, they're actually really easy to write though since I have refactored the codebase to allow for backends, here is npm: https://github.com/jdx/mise/commit/8d678e0ffb3b21a2874d6984a8d77e3e2372b4c7

I think the next problem might be how to deal with urls/repo names. I _think_ it'll work out of the box but that might require some (probably minor) modifications.

Also, within the codebase, I've settled on calling these "forges"—the idea being they stamp out instances of "tools". A "forge" is something like `asdf:node` or `npm:prettier` and a "tool" is that w/ a version: `asdf:node@20` and `npm:prettier@20`. This naming convention has actually been a problem since the beginning.

This was important in the codebase where "plugin" has been an overloaded term. These should not be thought of as plugins. In the future, a plugin may provide one or many forges or none at all in the case of a plugin for tasks and environments.

I have not made this change in anything user-facing (perhaps in some error messages) but that probably should be done at some point—though when that happens the change should be made everywhere at the same time for consistency. Part of the reason is I'm still a little undecided if "forge" should be the final name—and you all know how I can be about naming things :). It will make explaining mise a bit harder since right now there is a 1:1 mapping between plugins and forges but that will need to change.

**Comment by RingOfStorms** [comment]
Edit: Totally just assumed this is how it worked but jdx implemented it how I thought it should be already my bad for not trying it first.

Also for the `npm i -g` this bypasses one of my biggest issues with things installed with npm. Which is when I use _mise_ in projects that define different node versions the global install is only for that _specific_ version of node and I can lose a bunch of my installed tools just by moving around my directories in the command line unless I install them on every single node version managed by _mise_

It would be great for _mise_ to be the tool that makes behavior consistent and pins things in the path when installed via _mise_. So in other words if I installed a npm based tool using _mise_ I would expect it to ensure it is always on the path, regardless of which version of node is on my path at that time.

This is your tool so it is up to you, but I feel liek having _mise_ just call commands of other tools direcly like this with no extra handling isn't super useful, and may even may _mise_ more confusing to use. As a user who may only know 1-2 of the 20 possible backend tools I would have no idea what command I should run, esp when many may actually work but have large tradeoffs.

If I use `mise install npm:cspell:latest` and then ce into my typescript repo (with a different node version than where I ran the previous cspell install) and suddenly don't have cspell available on my path I'd be pretty confused. Then maybe I'd try to install it with something else `mise install eget:cspell` and now there are two installed in different places and one may be used instead of the other at weird times.

**Reply by jdx** [reply]
> So in other words if I installed a npm based tool using mise I would expect it to ensure it is always on the path, regardless of which version of node is on my path at that time.

that is how it works though

**Reply by RingOfStorms** [reply]
oh sweet, my bad I haven't actually tried it yet :sweat_smile: I will refrain until I actually use the new features here

**Reply by jdx** [reply]
it might have problems though. I talked about this above in the description. How well a CLI installed with node-18 runs under node-20 I can't say.

**Comment by mbhutton** [comment]
>  For v1, I'm simply going to call whichever cargo or npm is on PATH. That may be something you've defined in your .mise.toml. It may be something in your system, just depends on how it is configured. This does mean you will need to manually setup cargo/npm/whatever ahead of time.

When bootstrapping a new system and running `mise install`, would the language level plugins  (e.g. rust or python) always be installed and activated before doing operations with associated package manager (e.g. cargo or pipx)? 

As a concrete example, if I have this in global or project settings and run `mise install`, would the python version be installed before the pip package? 
```
[tools]
python = "3.11.0"
pipx=PyGitUp = "latest"
```

If homebrew/nix/apt are also in the mix, perhaps they should also precede backends like pipx/cargo etc, as that's how some users install pipx.

**Reply by jdx** [reply]
not currently, mise installs things in parallel so you'll need to explicitly run `mise install python` first. This could be fixed but it's not a trivial change.

**Comment by endigma** [comment]
Go is done as of #1470

**Reply by endigma** [reply]
@jdx if you could make a more ergonomic way to work on this other than a devcontainer I could probably add more of these, the biggest thing that makes me not want to do it is dealing with the build, clean, test, build, clean, test stuff all manually. 

Maybe the dev area could have some sort of sandbox for mise to run in via a different base directory than `~/.local/share/mise` so it doesn't interfere with systemwide?

Or just docker for tests, as discussed in the PR.

**Reply by jdx** [reply]
the linter/pre-commit hook uses mise now so that part should be easier. I have a branch for e2e tests in docker but it needs more work

**Reply by jdx** [reply]
@endigma the new docker stuff is in, seems to work really well (on my machine at least). Just run `mise run docker:e2e`. Let me know if you have any other papercuts.

**Reply by endigma** [reply]
![image](https://github.com/jdx/mise/assets/34945114/8d3c58f9-5b91-42d1-a4d3-5721f55703ae)
@jdx

**Reply by endigma** [reply]
not sure where $root is meant to be coming from

**Reply by endigma** [reply]
also, is there a way to run less than all of the e2e tests?

**Reply by jdx** [reply]
you should just be able to pass an argument to it

$root gets set here:

https://github.com/jdx/mise/blob/3412fa19e40ca66c4a1811a226b29804ed1f4d3b/src/cli/run.rs#L166

project_root must be null for you, though I don't know why that would be if you're running from the mise root...

**Reply by jdx** [reply]
oh I bet your mise is just out of date

**Reply by endigma** [reply]
out of date was the issue

**Comment by endigma** [comment]
Installing in a special path with a special cache, `pnpm`, `yarn`, `bun`/etc are likely useless, the benefit of using those would be sharing global node modules between them, and in `pnpm`'s case having the global symlink-store thing. They might be marginally faster on raw DL than npm though.

**Reply by jdx** [reply]
yeah I don't know who would need it, could maybe just be a setting for the npm backend

**Comment by endigma** [comment]
As far as I can tell, `nix`, `pnpm` and `eget` are DOA, no way to get them into the right paths or in `eget`'s case no way to list versions, although listing GitHub releases through the API might solve that. #1583 fixes the go forge to work properly.

**Comment by andrewthauer** [comment]
I think ruby gems would be a good one to add to this list as well. I think `gem` should have all the requirements to work properly with forge.

**Reply by jdx** [reply]
ruby doesn't look that hard, I found a way to list all the versions. It's a little awkward and there might be a better way, but this should function:

```sh
mise-docs  main ❯ gem search -aer neovim

*** REMOTE GEMS ***

neovim (0.9.1, 0.9.0, 0.8.1, 0.8.0, 0.7.1, 0.7.0, 0.6.2, 0.6.1, 0.6.0, 0.5.1, 0.5.0, 0.4.0, 0.3.3, 0.3.2, 0.3.1, 0.3.0, 0.2.5, 0.2.4, 0.2.3, 0.2.2, 0.2.1, 0.2.0, 0.1.0, 0.0.6, 0.0.5, 0.0.4, 0.0.3, 0.0.2, 0.0.1)
```

Now for installing, it looks like we'll need to add `--install-dir` and `--bindir` but I haven't tested this. It's important to make sure we install to the install path and don't install dependencies into the global gem location otherwise when you change ruby versions it will break.

There is also the question of how compatible ruby gems are when you change versions—which is likely going to be a problem for a lot of our tools.

**Reply by schmod** [reply]
Ruby would be great, because it'd let us version-control Cocoapods via mise.  (This is particularly notable because cocoapods just shipped two buggy releases in a row, and it would have saved a ton of headache for all of our devs to have been pinned to a non-buggy version)

**Reply by endigma** [reply]
Might be jenky to decide what ruby version to use for the tool?

**Reply by andrewthauer** [reply]
I was wondering how forge determines the runtime to use for installed packages. Looking at the npm backend, it seems the node_modules are vendored in the install location, but runs against whatever node version mise currently resolves to. I suspect this in general could result in unexpected and inconsistent behaviour. Unless I'm missing something in how it works?

**Reply by endigma** [reply]
It's probably possible to control which node version is used to install npm packages, but it's going to be very hard to set it up to *use* a specific one for a specific tool, it would require re-shimming as npm install-ed "executables" are just shims that call node anyway.

**Reply by jdx** [reply]
this is a known issue, I mentioned it in the initial description under "Runtime/compiler selection". Nobody has raised it yet but I'm sure it will be something that needs to be addressed and may require unique behavior in different languages.

**Reply by andrewthauer** [reply]
Looking at how homebrew handles this, it essentially creates a shim file with a special shebang pointing to the desired version of ruby, etc. to be executed. Not sure if that makes sense in `mise` or not though.

**Reply by andrewthauer** [reply]
I added a draft PR adding a `gem` backend here https://github.com/jdx/mise/pull/1657.

**Comment by eproxus** [comment]
A backend for Elixir's ~~`Mix.Install`~~[`mix escript.install`](https://hexdocs.pm/mix/Mix.Tasks.Escript.Install.html#content). would be nice.

**Reply by pepicrft** [reply]
@eproxus is `mix install` able to install Elixir CLIs globally? From what I understood there, `Mix.install` is a runtime API to run inside `iex`.

**Reply by eproxus** [reply]
My bad, I was thinking of [`mix escript.install`](https://hexdocs.pm/mix/Mix.Tasks.Escript.Install.html#content).

**Comment by BuildBackBuehler** [comment]
I'm glad to see _rtx_ going in the right direction, hot damn there's a lot going on since I was last in the loop! Homebrew is going in the other direction with the python changes...sigh (and here I am as a result)

1. I like the move away from plugins for the big-hitters. Plug-ins should be reserved for auxiliary accoutrements to those fellas. 

Backend is...ehhh. 

Framework is easily #1 to me.

Some others that come to mind: 
System, Base, Nexus, Design Entity (or Dgn. Sys.)

2. As for the meat & gravy, it would be nice to get Homebrew integrated into the rtx umbrella because the reason I'd gone back to Homebrew exclusively was because I could only use HB's basic dependencies if whatever "backend" was installed through HB. This was a problem because I couldn't use cmake/llvm/whatever to build some basic things I wanted to integrate into Ruby and/or Cargo. 

That is, if you/team doesn't want to slide those kinda dependencies into the chaos 😂 (^ above was definitely a Mac-only problem...IIRC if I could hardlink, I could've worked around it. I'm also not proficient let alone an expert in code world.)

3. Hope that Mise focuses on...

Cargo, Web PM (anything besides npm, but pref. a flat-packer, whatever is next-gen pnpm, Deno? Bun? IDK), Python/Pip(x) should be #1 priorities for sure! 

4. As for the great syntax debate -- I do believe that the dash `/` is too prevalent to be used for yet another delimiter. I love the whatever the duck `|` is called because it is probably the most convenient key out of all the delimiter characters at our disposal. But what about `>`? Makes sense with it being "This is Cargo's eza." I'm sure there's a reason they weren't mentioned I'm not privy to lol. 

Tanx for the amazing work you do JDX!

**Comment by florentulve** [comment]
HI,
I like the ergonomcis of mise but I would love a nix backend.

I am concerned about downloading binaries through third parties plugins.
Nix have a centralized package build script
Plus Nix have a colossal list of package.

Any help needed ? Would help if I can.

**Comment by cwegener** [comment]
I've just skimmed through this thread and I think I have a use case that has not been mentioned before.

I would like to be able to have something akin to what `pipx inject` does, where it allows me to inject more dependencies into a tool's existing managed storage location. https://realpython.com/python-pipx/#inject-dependencies-into-managed-environments

This is needed for tools that come with their own plugins (e.g. tailwindcss plugins, postcss plugins etc.)

I don't have time myself to investigate and implement. But I thought I'd leave the idea here.

**Reply by jdx** [reply]
how is that different than installing a package with the tool's dependency manager? I don't see why mise needs a command for this

**Reply by cwegener** [reply]
In the case of the `npm` backend, this would result in the storage location being managed by `mise`. I would then get the benefit of using `mise implode` to get back my disk space. Without mise, I would need to rely on the package manager to be smart enough to have such functionality.

I can see how this might not be the something that mise should be burdened with, but package managers also don't make these kind of 'disk space management' tasks very easy.

**Comment by dudicoco** [comment]
Did anyone use [spack](https://github.com/spack/spack) for system packages? Seems like they lock transitive dependencies which would make it a great candidate for cross-platform system packages/libraries backend (git, gcc etc.).

**Reply by andrewthauer** [reply]
I remember looking at `spack` a couple of years ago when I was researching reproducible environments/build tools. Seemed pretty niche in comparison to `nix` which seems to cover the same space and use cases. I could be wrong though.

**Comment by oussfyi** [comment]
Suggestion to add / transition to [`uv`](https://github.com/astral-sh/uv) as the default backend for Python installations (including tools). `uv` is is rapidly gaining popularity as a unified toolchain for the Python ecosystem, is very performant and reliable, and the company behind it will soon be taking over the [`python-build-standalone`](https://github.com/indygreg/python-build-standalone) project (read the announcement [here](https://astral.sh/blog/python-build-standalone)).

This will probably make managing Python (+ tools) installations easier and more reliable for end-users because it won't be necessary to compile CPython locally as it's the case currently -- which could fail completely, or worse: result in semi-broken / incomplete installation (see example below from my own machine). Also, `uv` being implemented in Rust would probably make it easier to integrate into `mise`.

@jdx  Please consider this suggestion, and thank you very much for all the great work you're doing on `mise`!

---
Example: A semi-broken Python installation
![image](https://github.com/user-attachments/assets/20fe3d15-5726-420a-a897-0adc8047f296)

**Comment by fakhrulhilal** [comment]
any chance to support brew? I think most of macos user using this one. For me, I don't have administrative privilege, so that could be an option for me. Certain package is available on brew, but not on every currently supported registry backend.

**Reply by risu729** [reply]
I think no. I don't remember the exact reason but jdx told me it's hard before. I guess it was basically because brew expects us to use latest version of tools.
It might be worth creating a backend plugin if you're interested.
https://mise.jdx.dev/backend-plugin-development.html

**Reply by jdx** [reply]
a brew backend would not be possible since brew doesn't support versions or portable binaries
